File: core/tool_chain_executor.py
"""Tool chain executor implementing the ReAct pattern for LLM tool calling.

Orchestrates the loop: LLM call -> parse tool_calls -> execute tools ->
feed results back -> repeat until final answer.

License: MIT
"""

import json
import time
from typing import Any

from flowforge.core import metrics
from flowforge.core.base_tool import ToolInput
from flowforge.core.tracing import get_logger

logger = get_logger("tool_chain_executor")


class ToolChainExecutor:
    """Orchestrates LLM + tool calls in a ReAct loop.

    The executor repeatedly calls the LLM with available tools, parses
    tool_calls from the response, executes them via the tool registry,
    and feeds results back until the LLM produces a final answer with
    no more tool_calls.

    Attributes:
        llm_client: The LLMClient instance for making LLM calls.
        tool_registry: The ToolRegistry for executing tool calls.
        event_bus: Optional event bus for emitting lifecycle events.
        max_iterations: Maximum number of LLM rounds before forcing a stop.
    """

    def __init__(self, llm_client, tool_registry, event_bus=None, max_iterations: int = 3):
        self.llm_client = llm_client
        self.tool_registry = tool_registry
        self.event_bus = event_bus
        self.max_iterations = max_iterations

    async def execute(
        self,
        task_id: str,
        messages: list[dict[str, Any]],
        tools: list[str] | None = None,
        system_prompt: str | None = None,
        model: str = "auto",
        persona: str = "default",
        agent_name: str = "helm_assistant",
        temperature: float = 0.7,
        max_tokens: int = 4000,
    ) -> dict[str, Any]:
        """Execute a tool chain loop.

        1. Build messages with system prompt + tool schemas
        2. Call LLM
        3. If LLM response contains tool_calls, execute them via tool_registry
        4. Add tool results to messages
        5. Repeat until no more tool_calls or max_iterations reached
        6. Return final response and execution trace

        Args:
            task_id: Task identifier for event emission.
            messages: Initial message list (OpenAI format).
            tools: Optional list of tool names to make available.
            system_prompt: Optional system prompt prepended to messages.
            model: Model hint ("auto" for automatic selection).
            persona: Persona identifier for model routing.
            agent_name: Agent name for model routing.
            temperature: LLM sampling temperature.
            max_tokens: Maximum tokens for LLM response.

        Returns:
            Dict with keys: content, execution_trace, iterations, total_tokens.
        """
        tool_schemas = self._build_tool_schemas(tools)

        all_messages: list[dict[str, Any]] = []
        if system_prompt:
            all_messages.append({"role": "system", "content": system_prompt})
        all_messages.extend(messages)

        execution_trace: list[dict[str, Any]] = []
        total_tokens = 0
        final_content = ""
        used_model = ""
        used_provider = ""
        tool_call_history: list[str] = []

        for iteration in range(self.max_iterations):
            if len(all_messages) > 12:
                all_messages = [all_messages[0]] + all_messages[-10:]
            self._emit_event(task_id, "tool_chain.iteration", {
                "iteration": iteration + 1,
                "max_iterations": self.max_iterations,
                "message_count": len(all_messages),
            })

            llm_params = {
                "messages": all_messages,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "stream": False,
                "persona": persona,
                "agent_name": agent_name,
                "task_id": task_id,
            }
            if tool_schemas:
                llm_params["tools"] = tool_schemas
            if model and model != "auto":
                llm_params["model"] = model

            tool_input = ToolInput(params=llm_params)
            tool_output = await self.llm_client.execute(tool_input)

            if tool_output.error:
                logger.error(f"LLM call failed in tool chain iteration {iteration + 1}: {tool_output.error}")
                return {
                    "content": final_content or f"LLM call failed: {tool_output.error}",
                    "execution_trace": execution_trace,
                    "iterations": iteration + 1,
                    "total_tokens": total_tokens,
                    "error": tool_output.error,
                }

            result = tool_output.result or {}
            content_text = result.get("content", "")
            tool_calls = result.get("tool_calls")
            raw_message = result.get("raw_message")
            tokens = result.get("tokens", 0)
            used_provider = result.get("provider", used_provider)
            used_model = result.get("model", used_model)
            total_tokens += tokens

            if not tool_calls:
                final_content = content_text
                self._emit_event(task_id, "tool_chain.complete", {
                    "iterations": iteration + 1,
                    "total_tokens": total_tokens,
                    "tool_calls_made": len(execution_trace),
                })
                return {
                    "content": final_content,
                    "execution_trace": execution_trace,
                    "iterations": iteration + 1,
                    "total_tokens": total_tokens,
                    "provider": used_provider,
                    "model": used_model,
                }

            assistant_msg = raw_message if raw_message else {"role": "assistant", "content": content_text}
            if "role" not in assistant_msg:
                assistant_msg["role"] = "assistant"
            all_messages.append(assistant_msg)

            for tool_call in tool_calls:
                func_info = tool_call.get("function", {})
                tool_name = func_info.get("name", "")
                arguments_str = func_info.get("arguments", "{}")
                tool_call_id = tool_call.get("id", "")

                try:
                    arguments = json.loads(arguments_str) if isinstance(arguments_str, str) else arguments_str
                except json.JSONDecodeError:
                    arguments = {}

                tool_call_history.append(tool_name)
                if len(tool_call_history) >= 3:
                    recent = tool_call_history[-3:]
                    if len(set(recent)) == 1:
                        logger.warning(f"Tool chain loop detected: {tool_name} called {len(tool_call_history)} times")
                        all_messages.append({
                            "role": "tool",
                            "tool_call_id": tool_call_id,
                            "content": json.dumps({"success": True, "result": "已完成，无需重复调用"}, ensure_ascii=False),
                        })
                        execution_trace.append({
                            "tool": tool_name,
                            "arguments": arguments,
                            "result": {"skipped": True, "reason": "loop_detected"},
                            "iteration": iteration + 1,
                        })
                        continue

                self._emit_event(task_id, "tool_chain.tool_call", {
                    "tool": tool_name,
                    "arguments": arguments,
                    "iteration": iteration + 1,
                })

                tool_result = await self._execute_tool_call(tool_name, arguments)

                all_messages.append({
                    "role": "tool",
                    "tool_call_id": tool_call_id,
                    "content": json.dumps(tool_result, ensure_ascii=False),
                })

                execution_trace.append({
                    "tool": tool_name,
                    "arguments": arguments,
                    "result": tool_result,
                    "iteration": iteration + 1,
                })

                self._emit_event(task_id, "tool_chain.tool_result", {
                    "tool": tool_name,
                    "iteration": iteration + 1,
                })

            final_content = content_text

        logger.warning(f"Tool chain reached max iterations ({self.max_iterations})")
        self._emit_event(task_id, "tool_chain.max_iterations", {
            "iterations": self.max_iterations,
            "total_tokens": total_tokens,
        })
        return {
            "content": final_content or "Max iterations reached without final answer",
            "execution_trace": execution_trace,
            "iterations": self.max_iterations,
            "total_tokens": total_tokens,
            "provider": used_provider,
            "model": used_model,
        }

    def _build_tool_schemas(self, tool_names: list[str] | None = None) -> list[dict[str, Any]]:
        schemas: list[dict[str, Any]] = []
        available_tools = self.tool_registry.list_tools()

        target_tools = tool_names if tool_names else available_tools

        for name in target_tools:
            if name == "llm":
                continue
            try:
                tool = self.tool_registry.get_tool(name)
            except Exception:
                continue

            func_schema = {
                "name": tool.name,
                "description": (tool.description or "")[:200],
                "parameters": tool.parameters_schema or {"type": "object", "properties": {}},
            }
            schemas.append({"type": "function", "function": func_schema})

        if len(schemas) > 10:
            schemas = schemas[:10]

        return schemas

    def _parse_tool_calls(self, llm_response: dict[str, Any]) -> list[dict[str, Any]]:
        """Extract tool calls from LLM response.

        Args:
            llm_response: The LLM response dict containing tool_calls.

        Returns:
            List of parsed tool call dicts with 'name' and 'arguments' keys.
        """
        tool_calls = llm_response.get("tool_calls", [])
        parsed = []
        for tc in tool_calls:
            func = tc.get("function", {})
            name = func.get("name", "")
            args_str = func.get("arguments", "{}")
            try:
                arguments = json.loads(args_str) if isinstance(args_str, str) else args_str
            except json.JSONDecodeError:
                arguments = {}
            parsed.append({"name": name, "arguments": arguments, "id": tc.get("id", "")})
        return parsed

    async def _execute_tool_call(self, tool_name: str, arguments: dict[str, Any]) -> dict[str, Any]:
        """Execute a single tool call and return the result.

        Args:
            tool_name: Name of the tool to execute.
            arguments: Arguments dict for the tool.

        Returns:
            Dict with 'success' flag and either 'result' or 'error'.
        """
        start = time.time()
        try:
            tool_input = ToolInput(params=arguments)
            tool_output = await self.tool_registry.execute(tool_name, tool_input)
            duration = time.time() - start
            metrics.record_tool_call(tool_name, duration)
            return {"success": True, "result": tool_output.result}
        except Exception as e:
            duration = time.time() - start
            metrics.record_tool_call(tool_name, duration)
            logger.error(f"Tool '{tool_name}' execution failed: {e}")
            return {"success": False, "error": str(e)}

    def _emit_event(self, task_id: str, event_type: str, payload: dict):
        if self.event_bus:
            self.event_bus.emit(task_id, event_type, payload)

File: packages/core/state/src/tool-chain-executor.ts
/**
 * tool-chain-executor — ReAct 工具链执行器（TS 重写自 `core/tool_chain_executor.py`，F27）。
 *
 * 编排循环：LLM 调用 → 解析 tool_calls → 执行工具 → 回填结果 → 重复直至
 * 最终答案或达到 max_iterations。内置消息窗口裁剪、同工具连续循环检测。
 *
 * @module @flowforge/core-state
 */

/** LLM 客户端最小接口（OpenAI 兼容消息格式）。 */
export interface LlmClientLike {
  execute(input: {
    params: Record<string, unknown>;
  }): Promise<{ result?: Record<string, unknown>; error?: string }>;
}

/** 工具描述（注册表 get_tool 返回）。 */
export interface ToolSchemaLike {
  name: string;
  description?: string;
  parameters_schema?: Record<string, unknown> | null;
}

/** 工具注册表最小接口。 */
export interface ToolRegistryLike {
  list_tools(): string[];
  get_tool(name: string): ToolSchemaLike;
  execute(
    name: string,
    input: { params: Record<string, unknown> },
  ): Promise<{ result?: unknown; error?: string }>;
}

/** 事件总线最小接口。 */
export interface EventBusLike {
  emit(taskId: string, eventType: string, payload: Record<string, unknown>): void;
}

/** 工具链执行结果。 */
export interface ToolChainResult {
  content: string;
  execution_trace: Array<{
    tool: string;
    arguments: Record<string, unknown>;
    result: Record<string, unknown>;
    iteration: number;
  }>;
  iterations: number;
  total_tokens: number;
  error?: string;
  provider?: string;
  model?: string;
}

/** OpenAI 格式消息。 */
export interface ChatMessage {
  role: string;
  content?: string;
  tool_call_id?: string;
}

/**
 * ReAct 工具链执行器 — 反复调用 LLM，解析 tool_calls 并经注册表执行，
 * 回填 tool 结果消息，直至 LLM 返回无 tool_calls 的最终答案。
 */
export class ToolChainExecutor {
  readonly llmClient: LlmClientLike;
  readonly toolRegistry: ToolRegistryLike;
  readonly eventBus: EventBusLike | undefined;
  readonly maxIterations: number;

  constructor(
    llmClient: LlmClientLike,
    toolRegistry: ToolRegistryLike,
    eventBus?: EventBusLike,
    maxIterations = 3,
  ) {
    this.llmClient = llmClient;
    this.toolRegistry = toolRegistry;
    this.eventBus = eventBus;
    this.maxIterations = maxIterations;
  }

  /**
   * 执行工具链循环。
   *
   * @param taskId 任务标识（事件发射用）。
   * @param messages 初始消息列表（OpenAI 格式）。
   * @param tools 可用工具名列表（缺省用注册表全部）。
   * @param systemPrompt 前置系统提示词。
   * @param model 模型提示（"auto" 自动选择）。
   * @param persona 模型路由 persona。
   * @param agentName 模型路由 agent 名。
   * @param temperature 采样温度。
   * @param maxTokens 响应最大 token。
   */
  async execute(options: {
    taskId: string;
    messages: ChatMessage[];
    tools?: string[];
    systemPrompt?: string;
    model?: string;
    persona?: string;
    agentName?: string;
    temperature?: number;
    maxTokens?: number;
  }): Promise<ToolChainResult> {
    const {
      taskId,
      messages,
      tools,
      systemPrompt,
      model = 'auto',
      persona = 'default',
      agentName = 'helm_assistant',
      temperature = 0.7,
      maxTokens = 4000,
    } = options;

    const toolSchemas = this.buildToolSchemas(tools);

    const allMessages: ChatMessage[] = [];
    if (systemPrompt !== undefined && systemPrompt !== '') {
      allMessages.push({ role: 'system', content: systemPrompt });
    }
    allMessages.push(...messages);

    const executionTrace: ToolChainResult['execution_trace'] = [];
    let totalTokens = 0;
    let finalContent = '';
    let usedModel = '';
    let usedProvider = '';
    const toolCallHistory: string[] = [];

    for (let iteration = 0; iteration < this.maxIterations; iteration += 1) {
      // 消息窗口裁剪：保留系统消息 + 最近 10 条
      if (allMessages.length > 12) {
        allMessages.splice(1, allMessages.length - 11);
      }
      this.emitEvent(taskId, 'tool_chain.iteration', {
        iteration: iteration + 1,
        max_iterations: this.maxIterations,
        message_count: allMessages.length,
      });

      const llmParams: Record<string, unknown> = {
        messages: allMessages,
        temperature,
        max_tokens: maxTokens,
        stream: false,
        persona,
        agent_name: agentName,
        task_id: taskId,
      };
      if (toolSchemas.length > 0) {
        llmParams['tools'] = toolSchemas;
      }
      if (model !== '' && model !== 'auto') {
        llmParams['model'] = model;
      }

      const toolOutput = await this.llmClient.execute({ params: llmParams });

      if (toolOutput.error) {
        return {
          content: finalContent || `LLM call failed: ${toolOutput.error}`,
          execution_trace: executionTrace,
          iterations: iteration + 1,
          total_tokens: totalTokens,
          error: toolOutput.error,
        };
      }

      const result = toolOutput.result ?? {};
      const contentText = String(result['content'] ?? '');
      const toolCalls = result['tool_calls'];
      const rawMessage = result['raw_message'];
      const tokens = Number(result['tokens'] ?? 0);
      usedProvider = String(result['provider'] ?? usedProvider);
      usedModel = String(result['model'] ?? usedModel);
      totalTokens += tokens;

      // 无 tool_calls → 最终答案
      if (!Array.isArray(toolCalls) || toolCalls.length === 0) {
        finalContent = contentText;
        this.emitEvent(taskId, 'tool_chain.complete', {
          iterations: iteration + 1,
          total_tokens: totalTokens,
          tool_calls_made: executionTrace.length,
        });
        return {
          content: finalContent,
          execution_trace: executionTrace,
          iterations: iteration + 1,
          total_tokens: totalTokens,
          provider: usedProvider,
          model: usedModel,
        };
      }

      // 回填 assistant 消息（优先 raw_message 保留原始 tool_calls）
      const assistantMsg =
        rawMessage !== null && typeof rawMessage === 'object'
          ? { ...(rawMessage as Record<string, unknown>) }
          : { role: 'assistant', content: contentText };
      if (!('role' in assistantMsg)) {
        (assistantMsg as Record<string, unknown>)['role'] = 'assistant';
      }
      allMessages.push(assistantMsg as ChatMessage);

      for (const toolCall of toolCalls as Array<Record<string, unknown>>) {
        const funcInfo =
          toolCall['function'] !== null && typeof toolCall['function'] === 'object'
            ? (toolCall['function'] as Record<string, unknown>)
            : {};
        const toolName = String(funcInfo['name'] ?? '');
        const argumentsStr = funcInfo['arguments'] ?? '{}';
        const toolCallId = String(toolCall['id'] ?? '');

        const arguments_ = parseArguments(argumentsStr);

        // 循环检测：同工具连续 3 次 → 跳过执行并回填提示
        toolCallHistory.push(toolName);
        if (toolCallHistory.length >= 3) {
          const recent = toolCallHistory.slice(-3);
          if (new Set(recent).size === 1) {
            allMessages.push({
              role: 'tool',
              tool_call_id: toolCallId,
              content: JSON.stringify({
                success: true,
                result: '已完成，无需重复调用',
              }),
            });
            executionTrace.push({
              tool: toolName,
              arguments: arguments_,
              result: { skipped: true, reason: 'loop_detected' },
              iteration: iteration + 1,
            });
            continue;
          }
        }

        this.emitEvent(taskId, 'tool_chain.tool_call', {
          tool: toolName,
          arguments: arguments_,
          iteration: iteration + 1,
        });

        const toolResult = await this.executeToolCall(toolName, arguments_);

        allMessages.push({
          role: 'tool',
          tool_call_id: toolCallId,
          content: JSON.stringify(toolResult),
        });

        executionTrace.push({
          tool: toolName,
          arguments: arguments_,
          result: toolResult,
          iteration: iteration + 1,
        });

        this.emitEvent(taskId, 'tool_chain.tool_result', {
          tool: toolName,
          iteration: iteration + 1,
        });
      }

      finalContent = contentText;
    }

    this.emitEvent(taskId, 'tool_chain.max_iterations', {
      iterations: this.maxIterations,
      total_tokens: totalTokens,
    });
    return {
      content: finalContent || 'Max iterations reached without final answer',
      execution_trace: executionTrace,
      iterations: this.maxIterations,
      total_tokens: totalTokens,
      provider: usedProvider,
      model: usedModel,
    };
  }

  /** 从注册表构建工具 schema 列表（跳过 llm，描述截断 200，最多 10 个）。 */
  buildToolSchemas(toolNames?: string[]): Array<Record<string, unknown>> {
    const schemas: Array<Record<string, unknown>> = [];
    let availableTools: string[] = [];
    try {
      availableTools = this.toolRegistry.list_tools();
    } catch {
      availableTools = [];
    }

    const targetTools = toolNames ?? availableTools;

    for (const name of targetTools) {
      if (name === 'llm') {
        continue;
      }
      let tool: ToolSchemaLike;
      try {
        tool = this.toolRegistry.get_tool(name);
      } catch {
        continue;
      }
      const funcSchema: Record<string, unknown> = {
        name: tool.name,
        description: (tool.description ?? '').slice(0, 200),
        parameters:
          tool.parameters_schema ?? { type: 'object', properties: {} },
      };
      schemas.push({ type: 'function', function: funcSchema });
    }

    if (schemas.length > 10) {
      schemas.splice(10);
    }
    return schemas;
  }

  /** 从 LLM 响应中解析 tool_calls（供外部复用）。 */
  parseToolCalls(llmResponse: Record<string, unknown>): Array<{
    name: string;
    arguments: Record<string, unknown>;
    id: string;
  }> {
    const toolCalls = llmResponse['tool_calls'];
    if (!Array.isArray(toolCalls)) {
      return [];
    }
    const parsed: Array<{
      name: string;
      arguments: Record<string, unknown>;
      id: string;
    }> = [];
    for (const tc of toolCalls) {
      const func =
        tc !== null && typeof tc === 'object'
          ? ((tc as Record<string, unknown>)['function'] as Record<string, unknown> | undefined)
          : undefined;
      const name = String(func?.['name'] ?? '');
      const argsStr = func?.['arguments'] ?? '{}';
      const id = String((tc as Record<string, unknown>)['id'] ?? '');
      parsed.push({ name, arguments: parseArguments(argsStr), id });
    }
    return parsed;
  }

  /** 执行单个工具调用，返回 {success, result|error}。 */
  private async executeToolCall(
    toolName: string,
    arguments_: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    try {
      const toolOutput = await this.toolRegistry.execute(toolName, {
        params: arguments_,
      });
      return { success: true, result: toolOutput.result ?? null };
    } catch (error) {
      return { success: false, error: String(error) };
    }
  }

  /** 发射生命周期事件（无事件总线时静默）。 */
  private emitEvent(
    taskId: string,
    eventType: string,
    payload: Record<string, unknown>,
  ): void {
    if (this.eventBus !== undefined) {
      try {
        this.eventBus.emit(taskId, eventType, payload);
      } catch {
        // 事件发射失败不影响主流程
      }
    }
  }
}

/** 解析工具参数字符串（JSON 失败回退空对象）。 */
function parseArguments(argumentsStr: unknown): Record<string, unknown> {
  if (typeof argumentsStr === 'string') {
    try {
      const parsed: unknown = JSON.parse(argumentsStr);
      if (parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)) {
        return parsed as Record<string, unknown>;
      }
      return {};
    } catch {
      return {};
    }
  }
  if (argumentsStr !== null && typeof argumentsStr === 'object' && !Array.isArray(argumentsStr)) {
    return argumentsStr as Record<string, unknown>;
  }
  return {};
}
